_posts/2019-01-21-vet-analyzer.md

```diff
@@ -0,0 +1,315 @@
+
+Let us peek into the [SelectorExpr](https://godoc.org/go/ast#SelectorExpr) to get these. A `SelectorExpr` again has two parts- `X` and `Sel`. If we take an example of `db.Exec()`- then `db` is `X`, and `Exec` is `Sel`. Matching the function name is easy. But to get the type info, we need to take help of `analysis.Pass` passed in the `run` function.
+
+[Pass](https://godoc.org/golang.org/x/tools/go/analysis#Pass) contains a `TypesInfo` field which contain type information about the package. We need to use that to get the type of `X` and verify that the object comes from the `database/sql` package and is of type `*sql.DB`.
+
+```go
+inspect.Preorder(nodeFilter, func(n ast.Node) {
+	call := n.(*ast.CallExpr)
+	sel, ok := call.Fun.(*ast.SelectorExpr)
+	if !ok {
+		return
+	}
+
+	// Get the type of X
+	typ, ok := pass.TypesInfo.Types[sel.X]
+	if !ok {
+		return
+	}
+
+	var nTyp *types.Named
+	switch t := typ.Type.(type) {
+	case *types.Pointer:
+		// If it is a pointer, get the element
+		nTyp = t.Elem().(*types.Named)
+	case *types.Named:
+		nTyp = t
+	}
+
+	if nTyp == nil {
+		return
+	}
+})
+```
+
+Now, from `nTyp` we can get the type info of `X` and directly match the function name from `Sel`.
+
+```go
+// Get the function name
+sel.Sel.Name // == Exec
+
+// Get the object name
+nTyp.Obj().Name() // == DB
+
+// Check the import of the object
+nTyp.Obj().Pkg().Path() // == database/sql
+```
+
+### Extract the query string
+
+Alright ! We have successfully filtered out only expressions of type `(*sql.DB).Exec`. The only thing remaining is to extract the query string from the `CallExpr` and check it for syntax errors.
+
+So far, we have been dealing with the `Fun` field of a `CallExpr`. To get the query string, we need to access `Args`. A `db.Exec` call will have the query string as its first param and the arguments follow after. We will get the first element of the `Args` slice and then use `TypesInfo.Types` again to get the value of the argument.
+
+```go
+// Code continues from before.
+
+arg0 := call.Args[0]
+typ, ok := pass.TypesInfo.Types[arg0]
+if !ok || typ.Value == nil {
+	return
+}
+
+_ = constant.StringVal(typ.Value) // Gives us the query string ! (constant is from "go/constant")
+```
+
+Note that this doesn't work if the query string is a variable. A lot of codebases have a query template string and generate the final query string dynamically. So for eg.
+
+```go
+q := `SELECT %s FROM certificates WHERE date=$1;`
+query := fmt.Sprintf(q, table)
+db.Exec(query, date)
+```
+
+will not work.
+
+All that is left is for us to check the query string for syntax errors. We will use the `github.com/lfittl/pg_query_go` package for that. And if we get an error, `pass` has a `Reportf` helper method to print out diagnostics found during a vet pass. So -
+
+```go
+query := constant.StringVal(typ.Value)
+_, err := pg_query.Parse(query)
+if err != nil {
+	pass.Reportf(call.Lparen, "Invalid query: %v", err)
+	return
+}
```

Comment:
Yeah, makes sense.

Reply:
I read it again. If I just mention a missed comma, I have to skip mentioning that the missing `$5` was something that was tripping me up often. And personally, I don't miss a comma, so it would be unnatural for me to write something that is not completely genuine.

I will add some extra text to clarify that this post just focuses on syntactic checks.